src/main/java/org/embulk/output/kintone/KintonePageOutput.java

```diff
@@ -185,183 +193,133 @@ private void updatePage(final Page page)
+                    // already validated in KintoneOutputPlugin.open()
+                    throw new RuntimeException("unreachable error");
+                }
+                ArrayList<Record> records = new ArrayList<>();
+                ArrayList<UpdateKey> updateKeys = new ArrayList<>();
+                pageReader.setPage(page);
 
-            if (!resp.hasNext()) {
-                break;
+                KintoneColumnVisitor visitor = new KintoneColumnVisitor(pageReader,
+                        task.getColumnOptions(),
+                        task.getUpdateKeyName().get());
+                while (pageReader.nextRecord()) {
+                    Record record = new Record();
+                    UpdateKey updateKey = new UpdateKey();
+                    visitor.setRecord(record);
+                    visitor.setUpdateKey(updateKey);
+                    for (Column column : pageReader.getSchema().getColumns()) {
+                        column.visit(visitor);
+                    }
+                    records.add(record);
+                    updateKeys.add(updateKey);
+
+                    if (records.size() == 10000) {
+                        upsert(records, updateKeys);
+                        records.clear();
+                        updateKeys.clear();
+                    }
+                }
+                if (records.size() > 0) {
+                    upsert(records, updateKeys);
+                }
             }
-        }
-        return allRecords;
+            catch (Exception e) {
+                throw new RuntimeException("kintone throw exception", e);
+            }
+        });
     }
 
-    abstract class UpsertPage<T>
+    private void upsert(ArrayList<Record> records, ArrayList<UpdateKey> updateKeys)
     {
-        public abstract List<T> getUpdateKeyValues(List<String> queryValues);
-        public abstract boolean existsRecord(List<T> updateKeyValues, Record record);
-
-        public void upsert(ArrayList<Record> records, ArrayList<UpdateKey> updateKeys)
-        {
-            if (records.size() != updateKeys.size()) {
-                throw new RuntimeException("records.size() != updateKeys.size()");
-            }
-
-            List<String> queryValues = updateKeys
-                .stream()
-                .map(k -> "\"" + k.getValue().toString() + "\"")
-                .collect(Collectors.toList());
-            List<T> updateKeyValues = getUpdateKeyValues(queryValues);
+        if (records.size() != updateKeys.size()) {
+            throw new RuntimeException("records.size() != updateKeys.size()");
+        }
 
-            ArrayList<Record> insertRecords = new ArrayList<>();
-            ArrayList<RecordForUpdate> updateRecords = new ArrayList<RecordForUpdate>();
-            for (int i = 0; i < records.size(); i++) {
-                Record record = records.get(i);
-                UpdateKey updateKey = updateKeys.get(i);
+        List<Record> distRecords = getRecordsByUpdateKey(updateKeys);
```

Comment:
kintone側にすでに存在する（=upsert対象の）レコードということなので、
```suggestion
        List<Record> existingRecords = getExiststingRecordsByUpdateKey(updateKeys);
```

がわかりやすいかなと思いました！